File: GitToolBox/src/main/java/zielu/gittoolbox/blame/calculator/CommitInfo.java
package zielu.gittoolbox.blame.calculator;

import com.intellij.openapi.vcs.history.VcsRevisionNumber;
import git4idea.GitRevisionNumber;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Objects;
import org.jetbrains.annotations.Nullable;

class CommitInfo {
  static final CommitInfo NULL = new CommitInfo(null);

  private final VcsRevisionNumber revisionNumber;
  private String authorName;
  private Instant authorTime;
  private ZoneOffset authorTimeOffset;
  private ZonedDateTime authorDateTime;
  private String summary;

  CommitInfo(String commitHash) {
    if (commitHash == null) {
      revisionNumber = VcsRevisionNumber.NULL;
    } else {
      revisionNumber = new GitRevisionNumber(commitHash);
    }
  }

  void setAuthorName(String authorName) {
    this.authorName = authorName;
  }

  void setAuthorTime(Instant authorTime) {
    this.authorTime = authorTime;
  }

  void setAuthorTimeOffset(ZoneOffset authorTimeOffset) {
    this.authorTimeOffset = authorTimeOffset;
  }

  void setSummary(String summary) {
    this.summary = summary;
  }

  VcsRevisionNumber getRevisionNumber() {
    return revisionNumber;
  }

  String getAuthorName() {
    return authorName;
  }

  @Nullable
  ZonedDateTime getAuthorDateTime() {
    if (authorTime == null) {
      return null;
    }
    if (authorDateTime == null) {
      ZoneOffset offset = authorTimeOffset == null ? ZoneOffset.UTC : authorTimeOffset;
      authorDateTime = ZonedDateTime.ofInstant(authorTime, offset);
    }
    return authorDateTime;
  }

  String getSummary() {
    return summary;
  }

  @Override
  public boolean equals(Object obj) {
    if (obj instanceof CommitInfo) {
      CommitInfo other = (CommitInfo) obj;
      return Objects.equals(this.revisionNumber, other.revisionNumber);
    }
    return false;
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/blame/calculator/CommitInfo.kt
package zielu.gittoolbox.blame.calculator

import com.intellij.openapi.vcs.history.VcsRevisionNumber
import git4idea.GitRevisionNumber
import java.time.Instant
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.ZonedDateTime

internal class CommitInfo private constructor(
  val revisionNumber: VcsRevisionNumber,
  val authorName: String? = null,
  val authorEmail: String? = null,
  val authorDateTime: ZonedDateTime? = null,
  val summary: String? = null
) {

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as CommitInfo

    if (revisionNumber != other.revisionNumber) return false

    return true
  }

  override fun hashCode(): Int = revisionNumber.hashCode()

  companion object {
    @JvmField
    val NULL: CommitInfo = CommitInfo(VcsRevisionNumber.NULL)
  }

  class Builder(val revisionNumber: VcsRevisionNumber) {
    var authorName: String? = null
      private set
    var authorEmail: String? = null
      private set
    var authorTime: Instant? = null
      private set
    var authorTimeOffset: ZoneId = ZoneOffset.UTC
      private set
    var summary: String? = null
      private set

    constructor(commitHash: String) : this(GitRevisionNumber(commitHash))

    fun authorName(authorName: String) = apply { this.authorName = authorName }

    fun authorEmail(authorEmail: String) = apply { this.authorEmail = authorEmail }

    fun authorTime(authorTime: Instant) = apply { this.authorTime = authorTime }

    fun authorTimeOffset(authorTimeOffset: ZoneId) = apply { this.authorTimeOffset = authorTimeOffset }

    fun summary(summary: String) = apply { this.summary = summary }

    private fun createAuthorDateTime(): ZonedDateTime? {
      if (authorTime != null) {
        return ZonedDateTime.ofInstant(authorTime, authorTimeOffset)
      }
      return null
    }

    fun build(): CommitInfo {
      val authorDateTime = createAuthorDateTime()
      return CommitInfo(revisionNumber, authorName, authorEmail, authorDateTime, summary)
    }
  }
}
